### ISSUE TITLE
[Bug]: HiDream-O1-Image generates pure noise on XPU — FlashAttention claims piecewise-span support on platforms that ignore spans

### ISSUE THREAD

**Issue by tthakkal** [vllm-project/vllm-omni#6882]
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code> (trimmed: CPU flag/vulnerability dump removed)</summary>

```text
==============================
        System Info
==============================
OS                           : Ubuntu 24.04.4 LTS (x86_64)
GCC version                  : (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
CMake version                : version 4.4.2
Libc version                 : glibc-2.39

==============================
       PyTorch Info
==============================
PyTorch version              : 2.13.0+xpu
Is debug build               : False
CUDA used to build PyTorch   : None
ROCM used to build PyTorch   : N/A

==============================
      Python Environment
==============================
Python version               : 3.12.3 (main, Jun 19 2026, 12:46:00) [GCC 13.3.0] (64-bit runtime)
Python platform              : Linux-6.14.0-1011-intel-x86_64-with-glibc2.39

==============================
       CUDA / GPU Info
==============================
Is CUDA available            : False

==============================
          CPU Info
==============================
Architecture                 : x86_64
Vendor ID                    : GenuineIntel
Core(s) per socket           : 48
Socket(s)                    : 2

==============================
          XPU Info
==============================
torch.xpu.is_available()     : True
torch.xpu.device_count()     : 2
  [0] Intel(R) Arc(TM) Pro B70 Graphics   total_memory=31.9 GiB   driver=1.15.39122+11
  [1] Intel(R) Arc(TM) Pro B70 Graphics   total_memory=31.9 GiB   driver=1.15.39122+11

==============================
Versions of relevant libraries
==============================
[pip3] numpy==2.2.6
[pip3] torch==2.13.0+xpu
[pip3] torchaudio==2.11.0+xpu
[pip3] torchvision==0.28.0+xpu
[pip3] transformers==5.14.1
[pip3] triton-xpu==3.7.2

==============================
         vLLM Info
==============================
vLLM Version                 : 0.28.0
vLLM-Omni Version            : dev (git sha: b81aeb7b)
vLLM Build Flags:
  CUDA Archs: Not Set; ROCm: Disabled

==============================
     Environment Variables
==============================
VLLM_TARGET_DEVICE=xpu
VLLM_WORKER_MULTIPROC_METHOD=spawn
VLLM_XPU_USE_SAMPLER_KERNEL=0
TORCHINDUCTOR_COMPILE_THREADS=1
```

</details>

### Your code version

<details>
<summary>The commit id or version of vllm</summary>

```text
0.28.0+xpu
```
</details>
<details>
<summary>The commit id or version of vllm-omni</summary>

```text
b81aeb7b86837f6fe8956f3aef83798ad26c5a26
```
</details>

### 🐛 Describe the bug

`HiDream-ai/HiDream-O1-Image` produces **pure noise** on XPU. There is no error, no warning, and the exit code is 0.

```bash
VLLM_OMNI_TARGET_DEVICE=xpu \
python examples/offline_inference/text_to_image/text_to_image.py \
  --model HiDream-ai/HiDream-O1-Image --tensor-parallel-size 2 \
  --prompt "A cat is sitting next to a sign that says 'HiDream-O1 vLLM-Omni'" \
  --height 2048 --width 2048 --num-inference-steps 50 --guidance-scale 5.0 --seed 42 \
  --output out.png
```

Scored against the CUDA-generated `tests/assets/hidream_o1/hidream_o1_baseline.png` with the repo's own `compute_image_ssim_psnr` helper (2048x2048 output LANCZOS-resized to the baseline's size, exactly as `tests/e2e/accuracy/test_hidream_o1_image.py` does):

| | SSIM | PSNR | Wall time |
|---|---|---|---|
| Observed on XPU | **0.0298** | **5.648 dB** | 195.9 s |
| Expected (thresholds in `test_hidream_o1_image.py`) | >= 0.8 | >= 19.0 | — |

The generation is also ~2.8x slower than it should be, which is the same symptom seen from the other side: the run is doing fully dense attention over the whole sequence instead of the structured attention the model asks for.

#### Root cause

`FlashAttentionBackend` declares span support unconditionally, as a plain class attribute:

```python
# vllm_omni/diffusion/attention/backends/flash_attn.py:24
class FlashAttentionBackend(AttentionBackend):
    supports_piecewise_spans: bool = True
```

But only `forward_cuda` actually dispatches `piecewise_attn` over `AttentionMetadata.full_attn_spans`. `forward_xpu` never reads the field — it looks only at `attn_metadata.attn_mask` and routes to `_forward_varlen_masked` or `_forward_varlen_dense`. The NPU forwards do not read spans either.

`hidream_o1_image_transformer.py` trusts that flag to decide whether it needs to build the equivalent 4D `attn_mask`:

```python
# vllm_omni/diffusion/models/hidream_o1_image/hidream_o1_image_transformer.py:518
all(
    layer.self_attn.attn.attn_backend.supports_piecewise_spans for layer in self.model.language_model.layers
)
```

On XPU that is `True`, so the model skips building the mask, and `forward_xpu` then sees `attn_mask=None` and runs **fully dense unmasked attention** — text rows that must stay causal attend to future image tokens. Nothing raises, because from each side's local point of view everything is consistent.

The flag describes the `forward_*` that actually runs, not the backend class, so a single class-level `True` cannot be correct for a backend with five platform forwards.

#### Two related dead guards

While tracing the above, two guards turned out to be unreachable for the same "flag is not what it looks like" reason:

1. `vllm_omni/diffusion/hooks/sequence_parallel.py:416` tests a bound method rather than calling it:

   ```python
   if not attn_backend.supports_attention_mask:   # always False -- this is a method object
   ```

   `mypy` already reports it on main: `error: Function "supports_attention_mask" could always be true in boolean context [truthy-function]`.

2. Four flash call sites use `torch.any(~attention_mask)` to detect padding (`flash_attn.py:445`, `flash_attn.py:492`, `flash_attn_hub.py:219`, `flash_attn_hub.py:386`). For a **bool** keep-mask that is right, but for an `int`/`uint8` keep-mask `~1 == -2`, which is truthy — so an all-valid int mask is unconditionally treated as padded and sent down the slower masked-varlen path. Reachable in-tree from Qwen-Image. The same expression silently inverts the meaning of an *additive* float mask, where `0.0` means valid.

#### Platforms affected

- **XPU** — noise, as above.
- **NPU** — same root cause, same silently-dense attention. Not verified (no NPU hardware available); reported from code reading only.
- **CUDA / ROCm / MUSA** — unaffected. `forward_hip` and `forward_musa` delegate to `forward_cuda`, which does read spans, so `True` is correct for exactly that family.

I have a fix that gates these capability flags per platform and will open a PR referencing this issue.

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.
